packages/build-tools/src/common/installDependencies.ts

```diff
@@ -1,24 +1,61 @@
 import path from 'path';
 
+import semver from 'semver';
 import { Job } from '@expo/eas-build-job';
 import spawn, { SpawnPromise, SpawnResult, SpawnOptions } from '@expo/turtle-spawn';
 
 import { BuildContext } from '../context';
 import { PackageManager, findPackagerRootDir } from '../utils/packageManager';
-import { isUsingYarn2 } from '../utils/project';
+import { isUsingModernYarnVersion } from '../utils/project';
 
 export async function installDependenciesAsync<TJob extends Job>(
   ctx: BuildContext<TJob>,
-  { logger, infoCallbackFn, cwd }: SpawnOptions
+  {
+    logger,
+    infoCallbackFn,
+    cwd,
+    sdkVersionFromPackageJson,
+    reactNativeVersionFromPackageJson,
+    withoutFrozenLockfile,
+  }: {
+    logger?: SpawnOptions['logger'];
+    infoCallbackFn?: SpawnOptions['infoCallbackFn'];
+    cwd?: SpawnOptions['cwd'];
+    sdkVersionFromPackageJson?: string;
+    reactNativeVersionFromPackageJson?: string;
+    withoutFrozenLockfile?: boolean;
+  }
 ): Promise<{ spawnPromise: SpawnPromise<SpawnResult> }> {
-  let args = ['install'];
-  if (ctx.packageManager === PackageManager.PNPM) {
-    args = ['install', '--no-frozen-lockfile'];
-  } else if (ctx.packageManager === PackageManager.YARN) {
-    const isYarn2 = await isUsingYarn2(ctx.getReactNativeProjectDirectory());
-    if (isYarn2) {
-      args = ['install', '--no-immutable', '--inline-builds'];
+  const shouldUseFrozenLockfile = Boolean(
+    !withoutFrozenLockfile &&
+      !ctx.env.EAS_NO_FROZEN_LOCKFILE &&
+      ((!!sdkVersionFromPackageJson && semver.satisfies(sdkVersionFromPackageJson, '>=52')) ||
+        (!!reactNativeVersionFromPackageJson &&
+          semver.satisfies(reactNativeVersionFromPackageJson, '>=0.76')))
+  );
```

Comment:
the logic looks complicated. maybe refactor some logic out to a function
```ts
  const shouldUseFrozenLockfile = Boolean(
    !withoutFrozenLockfile &&
    !ctx.env.EAS_NO_FROZEN_LOCKFILE) ?? useDefaultFrozenLockfile(packageJson);
    
  function useDefaultFrozenLockfile(packageJson: string): boolean {
    const sdkVersion = semver.coerce(packageJson?.dependencies?.expo)?.version;
    const reactNativeVersion = semver.coerce(
      packageJson?.dependencies?.['react-native']
    )?.version;
    if (!sdkVersion || sdkVersion.major < 52) {
      return false;
    }
    if (!reactNativeVersion || reactNativeVersion.major < 76) {
      return false;
    }
    return true;
  }
```

passing packageJson to the function and we can also save a little time without doing semver parsing when `EAS_NO_FROZEN_LOCKFILE=true`.